In [2]:
import os
os.chdir(r"C:\Users\zouff\PycharmProjects\Y2\Datascience2\data_aux")

import numpy as np
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
import math
import six
import sys
import scipy.stats as stats
from scipy.stats import binom
from scipy.stats import norm
from scipy.stats import poisson
from scipy.stats import t

pd.set_option('display.max_columns', None)
pd.set_option('display.max_row', None)

missing_values = ['n/a', 'na', 'nan', 'N/A', 'NA', 'NaN', 'NAN', '--', 'Missing', 'missing', 'Unknown', 'unknown', 'UNKNOWN']



# 1. EXAMPLE


In [3]:
df_c = pd.read_csv('consumptionLaptops.csv',decimal='.')
df_c.info()
x_bar = df_c.consumptionLaptops.mean()
print()
print('Sample mean : ' + str(x_bar))
s = df_c.consumptionLaptops.std()
print('Sample standard deviation : ' + str(s))
n = len(df_c)



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 1 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   consumptionLaptops  30 non-null     float64
dtypes: float64(1)
memory usage: 372.0 bytes

Sample mean : 39.99999999999999
Sample standard deviation : 19.999999999999996


What if we took a sample of the sample?

In [26]:
sample = df_c.sample(5)
x_bar_sample = sample.consumptionLaptops.mean()
print('Sample mean : ' + str(x_bar_sample))
s_sample = sample.consumptionLaptops.std()
print('Sample standard deviation : ' + str(s_sample))

Sample mean : 46.84288799172478
Sample standard deviation : 24.412124298963523


# 2. CONFIDENCE INTERVALS

## Normal distribution

2 standard deviations confidence interval:

In [31]:
low = x_bar - 2 * s / math.sqrt(n)
print('low : ' + str(low))
upper = x_bar + 2 * s / math.sqrt(n)
print('upper : ' + str(upper))

low : 32.69703256659778
upper : 47.30296743340221


In [35]:
P = 0.955
factor = norm.ppf((1+P)/2) 
print('factor : ' + str(factor))
interval = (x_bar - (factor * s / math.sqrt(n)),x_bar + (factor * s / math.sqrt(n)))
print(interval)

factor : 2.0046544617650968
(32.68003687525252, 47.31996312474747)


In [39]:
norm.interval(confidence=0.955,loc=x_bar, scale=s/math.sqrt(30))

(32.680036875252526, 47.31996312474747)

Until now, we worked with a confidence interval of 2 standard deviations, which is about 95.5% (factor = 2).

What if we want to use a 95% confidence interval?

In [42]:
P = 0.95
factor = norm.ppf((1+P)/2) 
print('factor : ' + str(factor))

interval = (x_bar - (factor * s / math.sqrt(n)),x_bar + (factor * s / math.sqrt(n)))
print(interval)

factor : 1.959963984540054
(32.84322342513136, 47.15677657486862)


In [8]:
alpha = 0.05
factor = norm.ppf(1-(alpha/2))
print('factor : ' + str(factor))

interval = (x_bar - (factor * s / math.sqrt(n)),x_bar + (factor * s / math.sqrt(n)))
print(interval)

factor : 1.959963984540054
(32.84322342513138, 47.15677657486864)


In [ ]:
norm.interval(confidence=0.95, loc=x_bar, scale=s/math.sqrt(n))

Cumulative distribution function versus percent point function

In [9]:
print(norm.cdf(1.644854))
print(norm.ppf(0.95))

0.950000038474587
1.6448536269514722


In [ ]:
print(norm.ppf(0.975))
print(norm.cdf(1.959963))

## Student's t-distribution

In [10]:
alpha = 0.05
factor = t.ppf(1-(alpha/2), n-1)
print('factor : ' + str(factor))

interval = (x_bar-factor*s/math.sqrt(n), x_bar+factor*s/math.sqrt(n))
print(interval)

factor : 2.045229642132703
(32.53187726483801, 47.468122735162005)


In [ ]:
t.interval(confidence=0.955,df=n-1,loc=x_bar,scale=s/math.sqrt(n))

In [11]:
alpha = 0.05
factor = t.ppf(1-(alpha/2), 5-+1)
print('factor : ' + str(factor))

interval = (x_bar-factor*s/math.sqrt(n), x_bar+factor*s/math.sqrt(n))
print(interval)

factor : 2.7764451051977987
(29.86185590805575, 50.13814409194426)


# 3. HYPOTHESIS TESTING

## t-test

In [13]:
# Hypothesis: real mean is 31

alpha = 0.05
factor = t.ppf(1-(alpha/2),n-1)
print('factor : ' + str(factor))

mu0=31
interval = (mu0-factor*s/math.sqrt(n), mu0+factor*s/math.sqrt(n))
print(interval)

factor : 2.045229642132703
(23.531877264838002, 38.468122735162)


Given an alpha of 0.5, reject the hypothesis that the real mean (the mean of the complete population) is 31, because if the real mean would be 31, there is a 95% confidence interval that the observed mean (the mean observed in the sample) would have to be between 23.5 and 38.5. As the observed mean is 40, which is outside the interval 23.5 - 38.5, reject the hypothesis that the real mean is 31.

Mind there is a 5% chance of a type 1 error, i.e. a 5% chance to reject the hypothesis that the real mean is 31 while it is true. I.e. if the real mean would indeed be 31, there is a 5% chance that the observed mean is outside the interval 23.5 - 38.5. So there is a 5% chance that the result of our test - reject that the mean is 31 - is wrong.

In [12]:
t.interval(confidence=0.95, df=n-1, loc=x_bar,scale=s/math.sqrt(n))

(32.53187726483801, 47.468122735162005)

Given an alpha of 0.5, and given that the observed mean is 40 (the mean observed in the sample), there is a 95% confidence interval that the real mean (the mean in the complete population) is between 32.5 and 47.5. As the hypothesis presumes a real mean of 31, we reject the hypothesis that the real mean is 31, because 31 is outside the interval 32.5 - 47.5.

Mind there is a 5% chance of a type 1 error, there is a 5% chance to observe a mean of 40 if the real mean is 31.

In [ ]:
# Hypothesis: real mean is 33

alpha = 0.05
factor = t.ppf(1-(alpha/2),n-1)
print('factor : ' + str(factor))

mu0=33
interval = (mu0-factor*s/math.sqrt(n), mu0+factor*s/math.sqrt(n))
print(interval)

Given an alpha of 0.5, accept the hypothesis that the real mean is 33 (the mean of the complete population), because given that the real mean would be 33, there is a 95% confidence interval that the observed mean (the mean observed in the sample) is between 25.5 and 40.5, and the observed mean is 40, which is in the interval 26.5 - 40.5.

In [ ]:
t.interval(confidence=0.95, df=n-1, loc=x_bar,scale=s/math.sqrt(n))

Or, given an observed mean of 40, there is a 95% confidence interval that the real mean is between 32.5 and 57.5, and the assumed real mean of 33 is within that interval.

Mind that we also might make a type II error, it might be that the real mean is not 33. We accept the hypothesis that the real mean is 33 because our data does not contradict it. But it still might be the real mean is not 33.

# 5. QUESTIONNAIRE

Load data:

In [ ]:
studenq = pd.read_csv('Questionnaire 21-22.csv', delimiter=';', decimal='.')
studenq.info()

## Question 1

1.a. Add a column to the dataframe and place in it the length of a person expressed in his shoe size (i.e. length divided by shoe size)

In [ ]:
studenq["Length in shoe size"] = studenq['Length'] / studenq['Shoe Size']

1.b. Determine the mean and standard deviation

In [ ]:
x_bar = studenq["Length in shoe size"].mean()
s = studenq["Length in shoe size"].std()
print(x_bar, s)

## Question 2

2\. Specify the confidence interval (&alpha;=5%) for the average ratio of length to shoe size

In [ ]:
alpha=0.05 # This means P is 95%
n = len(studenq["Length in shoe size"])
factor = t.ppf(1-(alpha/2),df=n-1)
interval = (x_bar - factor * s/math.sqrt(n), x_bar + factor * s/math.sqrt(n))
print(interval)

In [ ]:
t.interval(confidence=1-alpha,df=n-1,loc=x_bar,scale=s/math.sqrt(n))

## Question 3

3\. Someone claims that the ratio of length to shoe size of a person is equal to 4.2 with a standard deviation of 0.05. Based on the data from the questionnaire, can you agree with this (with &alpha; = 5%)? And with &alpha; = 2.5%?

In [ ]:
# Alpha = 5%

alpha=0.05
mu0=4.2
s0=0.05
factor = norm.ppf(1-(alpha/2))
interval = (mu0 - factor * s0/math.sqrt(n), mu0 + factor * s0/math.sqrt(n))
print(interval)

In [ ]:
norm.interval(confidence=1-alpha,loc=mu0,scale=s0/math.sqrt(n))

In [ ]:
# Alpha = 2.5%

alpha=0.025
mu0=4.2
factor = norm.ppf(1-(alpha/2))
interval = (mu0 - factor * s0/math.sqrt(n), mu0 + factor * s0/math.sqrt(n))
print(interval)

In [ ]:
norm.interval(confidence=1-alpha, loc=mu0, scale=s0/math.sqrt(n))